executor.go

```diff
@@ -0,0 +1,224 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/minio/confess/tests"
+	"github.com/minio/confess/utils"
+	"github.com/minio/minio-go/v7"
+	"github.com/minio/pkg/console"
+	"github.com/minio/pkg/ellipses"
+)
+
+var slashSeparator = "/"
+
+// Config represents the confess executor configuration.
+type Config struct {
+	Hosts      []string      `json:"hosts"`
+	AccessKey  string        `json:"accessKey"`
+	SecretKey  string        `json:"secretKey"`
+	Insecure   bool          `json:"insecure,omitempty"`
+	Region     string        `json:"region,omitempty"`
+	Signature  string        `json:"signature"`
+	Bucket     string        `json:"bucket"`
+	OutputFile string        `json:"outputFile"`
+	FailAfter  int           `json:"failAfter"`
+	Duration   time.Duration `json:"duration"`
+}
+
+// Test interface defines the test.
+type Test interface {
+	Name() string
+	Init(ctx context.Context, config tests.Config, stats *tests.Stats) error
+	PreRun(ctx context.Context) error
+	Run(ctx context.Context) error
+	PostRun(ctx context.Context) error
+}
+
+// Validate - validates the config provided.
+func (c Config) Validate() error {
+	if len(c.Hosts) == 0 {
+		return errors.New("empty hosts")
+	}
+	if c.AccessKey == "" || c.SecretKey == "" {
+		return errors.New("accessKey and secretKey must be set")
+	}
+	switch strings.ToUpper(c.Signature) {
+	case "S3V4", "S3V2":
+	default:
+		return errors.New("unknown signature method. please use s3v4 or s3v2")
+	}
+	if c.Bucket == "" {
+		return errors.New("bucket name should not be empty")
+	}
+	return nil
+}
+
+// Executor represents the test executor.
+type Executor struct {
+	Bucket               string
+	Clients              []*minio.Client
+	EnableVersionedTests bool
+	LogFile              string
+	Stats                *tests.Stats
+	Duration             time.Duration
+	FailAfter            int
+	sLock                sync.Mutex
+}
+
+// NewExecutor returns the instance of the type Executor.
+func NewExecutor(ctx context.Context, config Config) (*Executor, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+	var endpoints []string
+	for _, hostStr := range config.Hosts {
+		hosts := strings.Split(hostStr, ",")
+		for _, host := range hosts {
+			if len(host) == 0 {
+				continue
+			}
+			if !ellipses.HasEllipses(host) {
+				endpoints = append(endpoints, host)
+				continue
+			}
+			patterns, perr := ellipses.FindEllipsesPatterns(host)
+			if perr != nil {
+				console.Fatalln(fmt.Errorf("unable to parse input arg %s: %s", patterns, perr))
+			}
+			for _, lbls := range patterns.Expand() {
+				endpoints = append(endpoints, strings.Join(lbls, ""))
```

Comment:
Are you getting `rune` here to `strings.Join(lbls. "")`?

Reply:
It is []string here

(https://github.com/minio/pkg/blob/main/ellipses/ellipses.go#L131)